server/commands/write_authzmodel.go

```diff
@@ -29,137 +30,42 @@ func NewWriteAuthorizationModelCommand(
-			allRelations[relationName] = struct{}{}
-		}
+	if err := typeSystem.ValidateRelationRewrites(); err != nil {
+		return nil, serverErrors.InvalidAuthorizationModelInput(err)
 	}
 
-	for _, td := range tds {
-		for relationName, usersetRewrite := range td.GetRelations() {
-			err := isUsersetRewriteValid(allRelations, typeToRelations[td.GetType()], td.GetType(), relationName, usersetRewrite)
-			if err != nil {
-				return err
-			}
+	if schemaVersion == typesystem.SchemaVersion1_1 {
+		if err := typeSystem.ValidateRelationTypeRestrictions(); err != nil {
+			return nil, serverErrors.InvalidAuthorizationModelInput(err)
 		}
 	}
 
-	return nil
-}
-
-// isUsersetRewriteValid checks if a particular userset rewrite is valid. The first argument is all the relations in
-// the model, the second argument is the subset of relations on the type where the rewrite occurs.
-func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map[string]struct{}, objectType, relation string, usersetRewrite *openfgapb.Userset) error {
-	if usersetRewrite.GetUserset() == nil {
-		return serverErrors.EmptyRewrites(objectType, relation)
+	id, err := id.NewString()
+	if err != nil {
+		return nil, err
 	}
 
-	switch t := usersetRewrite.GetUserset().(type) {
-	case *openfgapb.Userset_ComputedUserset:
-		computedUserset := t.ComputedUserset.GetRelation()
-		if computedUserset == relation {
-			return serverErrors.CannotAllowMultipleReferencesToOneRelation
-		}
-		if _, ok := relationsOnType[computedUserset]; !ok {
-			return serverErrors.RelationNotFound(computedUserset, "", nil)
-		}
-	case *openfgapb.Userset_TupleToUserset:
-		tupleset := t.TupleToUserset.GetTupleset().GetRelation()
-		if _, ok := relationsOnType[tupleset]; !ok {
-			return serverErrors.RelationNotFound(tupleset, "", nil)
-		}
-
-		computedUserset := t.TupleToUserset.GetComputedUserset().GetRelation()
-		if _, ok := allRelations[computedUserset]; !ok {
-			return serverErrors.RelationNotFound(computedUserset, "", nil)
-		}
-	case *openfgapb.Userset_Union:
-		for _, child := range t.Union.GetChild() {
-			err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, child)
-			if err != nil {
-				return err
-			}
-		}
-	case *openfgapb.Userset_Intersection:
-		for _, child := range t.Intersection.GetChild() {
-			err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, child)
-			if err != nil {
-				return err
-			}
-		}
-	case *openfgapb.Userset_Difference:
-		err := isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Base)
-		if err != nil {
-			return err
-		}
-
-		err = isUsersetRewriteValid(allRelations, relationsOnType, objectType, relation, t.Difference.Subtract)
-		if err != nil {
-			return err
-		}
+	utils.LogDBStats(ctx, w.logger, "WriteAuthzModel", 0, 1)
+	if err := w.backend.WriteAuthorizationModel(ctx, req.GetStoreId(), id, typeSystem.GetTypeDefinitions()); err != nil {
```

Comment:
I want to do this in a follow up PR as it is quite a lot and it would be nice to clean up the interfaces as
```
WriteAuthorizationModel(ctx, req.GetStoreId(), id, schemaVersion, typeSystem.GetTypeDefinitions())
```
is getting a bit long. I had such a follow up PR before but it will need a lot of ❤️  now.